### PR TITLE
Shmelev Ivan / Lab 1 / opt 1

### PR DESCRIPTION
**Pull request by luchshiy-shmel**
"Напечатать имена всех классов и их поля
Пример:
struct MyClass {
  int a;
  float b;
};

---

MyClass
  |_ a
  |_ b"

### CODE REVIEW

File: clang/test/lab1/shmelev_ivan/tests.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+class foo1 {
+  int a;
+  double b;
+  char c;
+};
+// CHECK: foo1
+// CHECK-NEXT: |_a
+// CHECK-NEXT: |_b
+// CHECK-NEXT: |_c
```
Comment (m-ly4) [inline clang/test/lab1/shmelev_ivan/tests.cpp:12]: Please print also static fields.

File: clang/test/lab1/shmelev_ivan/tests.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+struct foo2 {
```
Comment (m-ly4) [inline clang/test/lab1/shmelev_ivan/tests.cpp:14]: Please print types of fields also.

File: clang/test/lab1/shmelev_ivan/tests.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+};
+// CHECK: foo2
+// CHECK-NEXT: |_a
+// CHECK-NEXT: |_b
+// CHECK-NEXT: |_c
+
+template <typename T> class foo3 { T var; };
+// CHECK: foo3
+// CHECK-NEXT: |_var
+
```
Comment (m-ly4) [inline clang/test/lab1/shmelev_ivan/tests.cpp:27]: Please add also test case:
```
template<> class foo3si<int>;
```

File: clang/labs/lab1/shmelev_ivan/ShmelevPrintNamesPlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  bool ParseArgs(const clang::CompilerInstance &compiler,
+                 const std::vector<std::string> &args) override {
+    for (const auto &arg : args) {
+      if (arg == "--help") {
+        llvm::outs() << "This plugin traverses the AST of the codebase and "
+                        "prints the name and fields of each class or struct\n";
+        displayHelp = true;
+        break;
```
Comment (aobolensk) [inline clang/labs/lab1/shmelev_ivan/ShmelevPrintNamesPlugin.cpp:61]: Please, add check for help message in the tests

File: clang/labs/lab1/shmelev_ivan/ShmelevPrintNamesPlugin.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/Support/raw_ostream.h"
+
+class ClassExplorer : public clang::RecursiveASTVisitor<ClassExplorer> {
+public:
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *recordDecl) {
+    if (!recordDecl->isClass() && !recordDecl->isStruct()) {
```
Comment (aobolensk) [inline clang/labs/lab1/shmelev_ivan/ShmelevPrintNamesPlugin.cpp:10]: Please, add check that strict also work in the tests

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**